### PR TITLE
[Messenger][DX] Allow stamps to be passed directly to MessageBusInterface::dispatch()

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none
| License       | MIT
| Doc PR        | TODO

Me again o/!

This proposal is *purely* for DX. With `DelayStamp`, the proposal of QueueNameStamp and future things like `AmqpRoutingKeyStamp`, stamps are becoming more common for end users to use. This changes how it looks to use them:

```php
// before
$bus->dispatch(new Envelope(new SendSmsNotification('Hi!'), new DelayStamp(10), new QueueNameStamp('low')));

// after
$bus->dispatch(new SendSmsNotification('Hi!'), [new DelayStamp(10), new QueueNameStamp('low')]);
```

It's definitely a BC break, which is allowed because the component is experimental, though it should be minimized. This BC break shouldn't be felt by most end users, as creating your own bus is an advanced use-case. Even if you decorated it, you'll get an obvious error.

### CODE REVIEW

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Component/Messenger/TraceableMessageBus.php [hunk 1]
```diff
@@ -40,7 +42,7 @@ public function dispatch($message): Envelope
         ];
 
         try {
-            return $this->decoratedBus->dispatch($message);
+            return $this->decoratedBus->dispatch($message, ...$stamps);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/TraceableMessageBus.php]: I think the extra stamps should be part of $context

Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/TraceableMessageBus.php]: maybe we should replace `$message instanceof Envelope ? $message : new Envelope($message);`
by `$envelope = Envelope::wrap($message, $stamps);` everywhere?

**Review by ogizanagi — APPROVED** [review]
The anonymous class implementing `MessageBusInterface` in `src/Symfony/Component/Messenger/Tests/Transport/AmqpExt/Fixtures/long_receiver.php` needs to be updated as well.

**Comment by sroze** [comment]
I find it confusing: my original assumption would have been that the other arguments are other messages, in case I want to dispatch multiple of them at the same time. I know it's been done this way on `Envelope::__constructor` but looking back, I'm not convinced. 

This is much more explicit:
```php
$bus->dispatch(
   new Envelope(new SendSmsNotification('Hi!'))
       ->with(new DelayStamp(10), new QueueNameStamp('low'))
);
```

Or if it's *really* important, let's make it an array so we identify easily that it should not be other messages?
```php
$bus->dispatch(
   new SendSmsNotification('Hi!'),
   [ new DelayStamp(10), new QueueNameStamp('low') ]
);
```

**Comment by weaverryan** [comment]
Hmm, fair point. Wrapping it in an array still looks great to me, and indeed, a bit more clear. I can change to use that if others agree.

**Comment by fabpot** [comment]
I agree that using an array. looks a bit better.

**Comment by sroze** [comment]
I suggest we change the `Envelope::__construct` 2nd argument as well.

**Comment by weaverryan** [comment]
This is ready to go:

* 2nd arg added to MessageBusInterface: `array $stamps = []`
* "2nd" arg on `Envelope` changed from `StampInterface ...$stamps` to `array $stamps = []`

File: src/Symfony/Component/Messenger/CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,12 @@ CHANGELOG
 4.3.0
 -----
 
+ * [BC BREAK]: The `Envelope::__construct()` signature changed:
+   you can no longer pass an unlimited number of stamps as the second,
+   third, fourth, arguments etc: stamps are now an array passed to the
+   second argument.
+ * [BC BREAK]: The `MessageBusInterface::dispatch()` signature changed:
+   a second argument `array $sstamps = []` was added.
```
Comment (sroze) [inline src/Symfony/Component/Messenger/CHANGELOG.md]:
```suggestion
   a second argument `array $stamps = []` was added.
```

File: src/Symfony/Component/Messenger/CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,12 @@ CHANGELOG
 4.3.0
 -----
 
+ * [BC BREAK]: The `Envelope::__construct()` signature changed:
+   you can no longer pass an unlimited number of stamps as the second,
+   third, fourth, arguments etc: stamps are now an array passed to the
+   second argument.
+ * [BC BREAK]: The `MessageBusInterface::dispatch()` signature changed:
```
Comment (sroze) [inline src/Symfony/Component/Messenger/CHANGELOG.md]:
```suggestion
 * [BC BREAK] The `MessageBusInterface::dispatch()` signature changed:
```

File: src/Symfony/Component/Messenger/CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,12 @@ CHANGELOG
 4.3.0
 -----
 
+ * [BC BREAK]: The `Envelope::__construct()` signature changed:
```
Comment (sroze) [inline src/Symfony/Component/Messenger/CHANGELOG.md]:
```suggestion
 * [BC BREAK] The `Envelope::__construct()` signature changed:
```

File: src/Symfony/Component/Messenger/Tests/MessageBusTest.php [hunk 5]
```diff
@@ -134,4 +136,16 @@ public function testThatAMiddlewareCanUpdateTheMessageWhileKeepingTheEnvelopeSta
...
         $bus->dispatch($envelope);
     }
+
+    public function testItAddsTheStamps()
+    {
+        $finalEnvelope = (new MessageBus())->dispatch(new \stdClass(), [new DelayStamp(5), new BusNameStamp('bar')]);
+        $this->assertCount(2, $finalEnvelope->all());
+    }
+
+    public function testItAddsTheStampsToEnvelop()
```
Comment (sroze) [inline src/Symfony/Component/Messenger/Tests/MessageBusTest.php]:
```suggestion
    public function testItAddsTheStampsToEnvelope()
```

File: src/Symfony/Component/Messenger/TraceableMessageBus.php [hunk 6]
```diff
@@ -29,9 +29,10 @@ public function __construct(MessageBusInterface $decoratedBus)
     /**
      * {@inheritdoc}
      */
-    public function dispatch($message): Envelope
+    public function dispatch($message, array $stamps = []): Envelope
     {
         $envelope = $message instanceof Envelope ? $message : new Envelope($message);
+        $envelope = $envelope->with(...$stamps);
```
Comment (sroze) [inline src/Symfony/Component/Messenger/TraceableMessageBus.php]: That's where the `Envelope::wrap($messageOrEnvelope, array $stamps)` method would be convenient so we don't need to duplicate these two lines across files.

**Review by sroze — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Messenger/Envelope.php [hunk 7]
```diff
@@ -27,8 +27,9 @@ final class Envelope
 
     /**
      * @param object $message
+     * @param StampInterface[]
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Envelope.php]: missing argument name: $stamps

File: src/Symfony/Component/Messenger/Envelope.php [hunk 8]
```diff
@@ -40,6 +41,18 @@ public function __construct($message, StampInterface ...$stamps)
         }
     }
 
+    /**
+     * Makes sure the message is in an Envelope and adds the given stamps.
+     *
+     * @param object|Envelope $message
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Envelope.php]: missing docblock for $stamps

File: src/Symfony/Component/Messenger/MessageBusInterface.php [hunk 9]
```diff
@@ -22,6 +24,7 @@ interface MessageBusInterface
      * Dispatches the given message.
      *
      * @param object|Envelope $message The message or the message pre-wrapped in an envelope
+     * @param StampInterface[]
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/MessageBusInterface.php]: $stamps

**Comment by weaverryan** [comment]
Good catches - comments addressed!

**Comment by weaverryan** [comment]
Rebased & ready again

**Review by fabpot — APPROVED** [review]

**Review by sroze — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @weaverryan.

**Comment by sroze** [comment]
🤞 fixed failing tests in https://github.com/symfony/symfony/commit/fd4146c7682f3256d01282efbd62fa008b65e765.

**Comment by timiTao** [comment]
@weaverryan  I want to have clear bus interface, that clear purpose is to send a message. This configuration shows what message is sent and how it should be treated. Implementing this will force my class to 2 things - create the message and send them. `MessageBusInterface` is made more robust due to that.
This loses abstraction that `Messenger` was to give to the application - I should not concern about how the message is sent, but if it correctly accepted by bus and eventually delivered.

"preoccupied with whether or not they could, they didn't stop to think if they should." 

I think the original way was showing the better purpose of this functionality. Now is a blurred.
I'm interested, your opinion about design this interface in the context of abstraction.

**Comment by weaverryan** [comment]
Hi!

I don’t quite understand. We added an optional argument to the interface. Are you creating a class that implements this interface and the new argument causes you a problem? Or does the new optional argument cause an issue when you are using the interface?

Cheers!

**Comment by ro0NL** [comment]
if anything; we have 2 ways to do the same thing, as shown by the PR description. Not sure it's desired either :confused: 

currently messenger enforces a `Stamp` concept, which before was an implementation detail of the message.

**Comment by timiTao** [comment]
@weaverryan
 @ro0NL described this very good. I'm not talking about a technical problem, but design philosophy. 
I think, enriching `MessageBusInterface` interface with the `Stamp` concept unnecessary add more details. 
In my opinion, this improvement makes less explicit `MessageBusInterface` interface. Choices like that make the project depend on Messenger less supportable.
That is my opinion. KISS ;)

For `Stamps` we have:
> we have 2 ways to do the same thing

And for the general message, we have 3 ways to send it with `Stamp`:
###  message
```
$bus->dispatch(
   new SendSmsNotification('Hi!')
);
```
### message + array stamps
```
$bus->dispatch(
   new SendSmsNotification('Hi!'),
   [new DelayStamp(10), new QueueNameStamp('low')]
);
```
### envelope + array stamps

```
$bus->dispatch(
   new Envelope(new SendSmsNotification('Hi!')),
   [new DelayStamp(10), new QueueNameStamp('low')]
);
```
### envelope with stamps

```
$bus->dispatch(
   new Envelope(new SendSmsNotification('Hi!'))
       ->with(new DelayStamp(10), new QueueNameStamp('low'))
);
```

**Comment by nicolas-grekas** [comment]
`Envelope` has always been part of the "dispatch" contracts. Since `StampInterface` is an integral part of `Envelope`, stamps have always been part of the "dispatch" contracts too.
If there should only be one way, it could be to remove passing an Envelope directly.
But that would only create more WTF to me (e.g. pass an envelope and have it be considered the message?) + unneeded BC break maybe?

**Comment by ro0NL** [comment]
the nice thing is `Envelope extends object`, so we can see any message as type object, not knowing about evenelops/stamps.

I've no real issue with the change. Personally i like `dispatch(object $message): Envelope` (one can discard the return value, considering it void), it's straightforward.

Comparing this to a real life scenario; can we pass a message/envelop and its stamps separate to a mail deliverer?

**Comment by weaverryan** [comment]
> Comparing this to a real life scenario; can we pass a message/envelop and its stamps separate to a mail deliverer? Wouldnt they expect a stamped message/envelop instead?

Ha! Indeed. But what I would personally *really* like is for the mail deliverer to come to my house, put the items in the envelope for me, close it, put the stamps on, then take it to the post office. That's a bit outside of the scope of what their job should be perhaps, but it would be awesome! Even in the real world scenario, there's a balance between purity and having a great customer experience (neither should dominate).

**Comment by timiTao** [comment]
I really like your example, but I see this very straightforward.

For a simple message:
```
I would personally really like is for the mail deliverer to come to my house, put the items in the 
envelope for me, close it, put the stamps on, then take it to the post office.
```
you have:
```
$bus->dispatch(
   new SendSmsNotification('Hi!')
);
```

If you want advanced use, then you simply use `Envelope` with `Stamp`'s.

This change simply provides something between that is neither of these solutions and don't give use real advantage - only more blurred interface:
```
I would personally really like is for the mail deliverer to come to my house, put the items in the 
envelope for me, close it, then *maybe I will inform him about my stamps and he can add his stamps on*, then take it to the post office.
```

@nicolas-grekas I always like `Messenger` not to be strict and enforce own `Interface` over message. Then any part of `Envelope` should be also added to this interface?  Then why not add a third optional parameter, to act like `Envelope` function `withoutAll`? I think functionality/manage of `Stamp`'s should be closed only in `Envelope`.

**Comment by ro0NL** [comment]
> the mail deliverer to come to my house, put the items in the envelope for me, close it, put the stamps on, then take it to the post office. That's a bit outside of the scope of what their job should be perhaps

exactly, it shows multiple concerns. Not an issue per se, but has its cost. In real life this is called [Franking](https://en.wikipedia.org/wiki/Franking).

This also comes with a trust issue, some trust this to go right, others dont and stamp themselves.  Hence i think this leads to diversity.

The question is; if we pass an "envelop/message" and "stamps" separate, can we assume it will be franked implicitly? Im not sure it's obvious.

**Comment by nicolas-grekas** [comment]
Honestly, I lost track of what is discussed here.
If we're discussion variadic vs array as trailing argument, both are essentially the same and it's a matter of code style preference.

**Comment by ro0NL** [comment]
the "design philosophy" itself still :)

**Comment by nicolas-grekas** [comment]
it always has been on the dispatching side, so this is old :)

**Comment by ro0NL** [comment]
before we didnt stamp in dispatch() at least 

```diff
-        $envelope = $message instanceof Envelope ? $message : new Envelope($message);
+        $envelope = Envelope::wrap($message, $stamps);
```

**Comment by nicolas-grekas** [comment]
Semantically, this is strictly the same. There is a bijection between both styles:

`$bus->dispatch($msg, $stamps);`
vs
`$bus->dispatch(new Envelope($msg, $stamps));`

I see why the 1st style has been added. There is no technical reason not doto it.

File: src/Symfony/Component/Messenger/Envelope.php [hunk 10]
```diff
@@ -40,6 +41,19 @@ public function __construct($message, StampInterface ...$stamps)
         }
     }
 
+    /**
+     * Makes sure the message is in an Envelope and adds the given stamps.
+     *
+     * @param object|Envelope  $message
+     * @param StampInterface[] $stamps
+     */
+    public static function wrap($message, array $stamps = []): self
```
Comment (ro0NL) [inline src/Symfony/Component/Messenger/Envelope.php]: @weaverryan wouldnt stamp() be the better name? In case of explicit public usage ...

Comment (ro0NL) [reply src/Symfony/Component/Messenger/Envelope.php]: to clarify; if wrapping means "put it in an envelop" this method doesnt always "wrap", whereas `new Envelope($anything)` does.

This method wraps if needed, but always stamps.

Comment (weaverryan) [reply src/Symfony/Component/Messenger/Envelope.php]: So, are you suggesting a change? Or is it clear for you now?

Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Envelope.php]: It stamps only if stamps are provided.
The main use case is removing the "if" that the method contains from userland boilerplate.
Tldr, wrap() fits my mind :)

Comment (ro0NL) [reply src/Symfony/Component/Messenger/Envelope.php]: yeah it might fit; "we wrap it with stamps" :/ but to me that's "stamping", not "wrapping".

Just curious if wrap() and __construct() could be considered confusing...

Comment (ro0NL) [reply src/Symfony/Component/Messenger/Envelope.php]: > It stamps only if stamps are provided.

hm indeed, missed this detail :}  OK this is technical details are hard to convey with naming :) let's keep as is :+1:

**Comment by timiTao** [comment]
@nicolas-grekas
I simply wanted to show that the author took the direction that in my opinion, give more bad that good outcome.

 If you insist that they are the only use cases, then that you are correct. This would be correct. The original:
```
$bus->dispatch($msg);
```
and 
```
$bus->dispatch(new Envelope($msg, $stamps));
```
Semantically, this is strictly the not same. This led to ignoring the first use case. As mentioned [here](https://github.com/symfony/symfony/pull/30707#issuecomment-491067885), I gave you my arguments. Is there any argument, that could convince you? Maybe, but not mine. **If my arguments didn't convince you, my job here is done.**
